kernel/src/actions/deletion_vector.rs

```diff
@@ -206,16 +268,40 @@ impl DeletionVectorDescriptor {
                 );
 
                 // get the Bytes back out and limit it to dv_size
-                let position = cursor.position();
-                let mut bytes = cursor.into_inner();
-                let truncate_pos = position + dv_size as u64;
-                assert!(
-                    truncate_pos <= usize::MAX as u64,
-                    "Can't truncate as truncate_pos is > usize::MAX"
-                );
-                bytes.truncate(truncate_pos as usize);
-                let mut cursor = Cursor::new(bytes);
-                cursor.set_position(position);
+                let position: usize = cursor.position().try_into().map_err(|_| {
+                    Error::DeletionVector("position doesn't fit in usize".to_string())
+                })?;
+                let bytes = cursor.into_inner();
+                // -4 to remove CRC portion.
+                let dv_size_usize: usize = dv_size.try_into().map_err(|_| {
+                    Error::DeletionVector("dv_size doesn't fit in usize".to_string())
+                })?;
+                let truncate_pos = position + dv_size_usize - 4;
+
+                if validate_crc {
+                    require!(
+                        bytes.len() >= truncate_pos + 4,
+                        Error::DeletionVector(format!(
+                            "Can't validate CRC as there are not enough bytes {} < {}",
+                            bytes.len(),
+                            truncate_pos + 4
+                        ))
+                    );
+                    let mut crc_cursor: Cursor<Bytes> =
+                        Cursor::new(bytes.slice(truncate_pos..truncate_pos + 4));
+                    let crc = read_u32(&mut crc_cursor, Endian::Big)?;
+                    let crc32 = create_dv_crc32();
+                    // -4 to include magic portion of the CRC
+                    let expected_crc = crc32.checksum(&bytes.slice(position - 4..truncate_pos));
+                    require!(
+                        crc == expected_crc,
+                        Error::DeletionVector(format!(
+                            "CRC32 checksum mismatch. Got: {crc}, expected: {expected_crc}"
```

Comment:
To make it easier to debug:
```suggestion
                            "CRC32 checksum mismatch for {self.path}. Got: {crc}, expected: {expected_crc}"
```